feature(java-21): use ZGC as default Garbage Collector

CodeLieutenant opened the pull request:
1. JVM Tweaks for Java 21 with new Generational GC (ZGC) enabled by default

Argus Run: https://argus.scylladb.com/tests/scylla-cluster-tests/7c6a9e0e-d493-4ef2-8aa3-0364eab4775a


MUST BE MERGED AFTER #74 (by source code it does not depend on it, but it will fully validate if all flags are valid, and simple test run in CI will show us impact of this change)

Commit:
feature(java-21): use ZGC by default on java 21

Signed-off-by: Dusan Malusev <dusan.malusev@scylladb.dev>

Commit:
feature(java-21): use ZGC by default on java 21

Signed-off-by: Dusan Malusev <dusan.malusev@scylladb.dev>

CodeLieutenant commented:
> @CodeLieutenant do we have the result side by side, with and without this flag ? do we see any difference ?

I don't have two runs, any other run from before can be compared to this one.
As far as i see it, it has 2% improvement on top of 2-4% improvement for Java 21, and latency is a bit lower (which increases the throughput), aside from that, there is not much of a difference

fruch commented:
> > @CodeLieutenant do we have the result side by side, with and without this flag ? do we see any difference ?
> 
> @fruch I don't have two runs, any other run from before can be compared to this one.
> As far as i see it, it has 2% improvement on top of 2-4% improvement for Java 21, and latency is a bit lower (which increases the throughput), aside from that, there is not much of a difference.
> 
> Also it starts a bit faster then on Java 11, haven't check if the flags are the cause or just Java 21

This chance is for improving performance, hence I would live to see it in the context of the performance test in SCT, before we merge it.

I don't want surprises in perf tests, we have plenty of them already.

mykaul commented:
@CodeLieutenant - any news for this?

fruch commented:
## Review: JVM flags analysis

The core idea (ZGC for Java 21+) is sound — sub-ms pauses mean measured latency reflects server behavior, not client GC noise. However, most of the flags need rework.

### Flag-by-flag assessment

| Flag | Verdict | Reasoning |
|------|---------|-----------|
| `-XX:+UseZGC` | ✅ Keep | Good fit for a load generator. ~5-15% throughput trade-off vs G1 is acceptable for eliminating GC pause variance. |
| `-XX:+EnableDynamicAgentLoading` | ❌ Remove | Default is already `true` in Java 21. Redundant. Will be disallowed by default in Java 25+. |
| `-XX:+RegisterFinalizersAtInit` | ❌ Remove | Default is already `true`. Deprecated in JDK 22, will be removed. No reason to set. |
| `-XX:+UseStringDeduplication` | ❌ Remove — **harmful** | Breaks young-gen collection in Generational ZGC — keeps short-lived strings alive via weak handles, causing allocation stalls ([JDK-8347337](https://bugs.openjdk.org/browse/JDK-8347337)). No benefit for an ephemeral stress tool. |
| `-XX:MaxGCPauseMillis=20` | ❌ Remove | G1GC-specific knob. ZGC ignores it entirely — its pauses are sub-ms by design. |
| `-XX:+AlwaysPreTouch` | ⚠️ Optional | Adds startup cost to pre-fault heap pages. Minor benefit for long runs. Not harmful, not essential. |
| `-XX:ZUncommitDelay=5` | ❌ Remove | Way too aggressive (default is 300s). Under sustained stress load this causes constant uncommit/recommit cycles → page-fault latency spikes. |
| `-XX:+UseNUMA` | ⚠️ Optional | Auto-detected on most systems. ~2-5% benefit on multi-socket. Harmless. |
| `-XX:+UseLargePages` | ❌ Remove | Requires OS-level huge page pre-allocation. Fails silently or fatally without it. Not worth the operational complexity for a stress tool. |
| `-XX:+UseTransparentHugePages` | ❌ Remove | Causes unpredictable latency jitter from kernel page promotion. [Oracle ZGC docs](https://docs.oracle.com/en/java/javase/21/gctuning/z-garbage-collector.html) explicitly say "not recommended for latency-sensitive apps." Also contradicts having `UseLargePages`. |

### Missing critical flag

The PR description says "Generational GC (ZGC) enabled by default" — this is **incorrect for Java 21**. Generational ZGC requires an explicit flag until Java 23 ([JEP 439](https://openjdk.org/jeps/439), [JEP 474](https://openjdk.org/jeps/474)):

```
-XX:+ZGenerational    # Required for Java 21-22, default in Java 23+
```

Without it you get non-generational ZGC, which has ~4x worse throughput on Cassandra-like workloads per JEP 439 benchmarks.

### Cosmetic noise

~50% of the diff is re-indentation of the `case` block (2-space vs 4-space). This makes the functional change harder to review and creates unnecessary merge conflicts. Should be a separate commit or dropped.

### Suggested minimal version

```bash
if [ "$jvmver" -gt "20" ]; then
  JVM_OPTS="$JVM_OPTS \
    -XX:+UseZGC \
    -XX:+ZGenerational \
    -XX:+AlwaysPreTouch \
  "
fi
```

CodeLieutenant committed:
feature(java-21): tune ZGC flags and inline JVM options

Remove jvm-clients.options and inline JPMS flags directly into
bin/cassandra-stress. Replace harmful/redundant JVM flags with a
vetted set based on review feedback:

- Enable Generational ZGC (-XX:+ZGenerational) required for Java 21-22
- Remove flags harmful to ZGC (UseStringDeduplication, MaxGCPauseMillis,
  ZUncommitDelay=5, UseLargePages, UseTransparentHugePages)
- Remove redundant/deprecated flags (EnableDynamicAgentLoading,
  RegisterFinalizersAtInit)
- Add performance flags: disable tiered compilation, reduce thread stack,
  optimize string concat, parallel ref processing, autobox cache extension
- Pin heap size (equal Xms/Xmx) and disable ZUncommit to eliminate
  page-fault latency under sustained load
- All flags are overridable via CASSANDRA_STRESS_* environment variables

soyacz commented:
> @fruch @soyacz Should we make these changes the default in C-S? Since SCT perf tests got same flags (less then here, since I've added couple more that will improve performance)

I think yes. People that use c-s for demos, benchmarks outside of sct could benefit from more stable results

fruch commented on bin/cassandra-stress:
```diff
@@ -90,65 +89,175 @@ CLASSPATH="${CLASSPATH}:$(find_all_jars "$CASSANDRA_STRESS_HOME/lib")"
+
+# --- Heap Configuration ---
+# Override: CASSANDRA_STRESS_HEAP_SIZE (default: 8g)
+CASSANDRA_STRESS_HEAP_SIZE="${CASSANDRA_STRESS_HEAP_SIZE:-8g}"
```
this flag for each parameter, seems a bit excessive

not sure anyone would be using it

also it might be breaking the possibility to just overriding the  JVM_OPTS directly

fruch replied:
need to document it somewhere, of the available ones and such.

CodeLieutenant replied:
It will be fine as long as we use $JVM_OPTS at the end of everything so that I can override

fruch replied:
extra finding the 8g default can bite us, if the user is selecting 8gb loader, it would break once he'll run more then one command

we should be extra careful here, otherwise it would break lot of SCT logic
